webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/api/DataStructures.scala

```diff
@@ -58,53 +57,102 @@ final case class ApiAccountId(value: String) extends AnyVal
+  }
+
+  def exposeSecret(): String = {
+    secret
+  }
+
+  def hash(): ApiTokenHash = {
+    ApiTokenHash.hash(this)
+  }
+}
+
+object ApiTokenSecret {
+  private val tokenSize = 32
+
+  def generate(tokenGenerator: TokenGenerator, suffix: String = ""): ApiTokenSecret = {
+    val completeSuffix = if (suffix.isEmpty) "" else "-" + suffix
+    ApiTokenSecret(tokenGenerator.newToken(tokenSize) + completeSuffix)
+  }
+}
+
+/*
+ * There are two versions of token hashes:
  *
- * * v1: 32 alphanumeric characters stored as clear text
- *       they are also displayed in clear text in the interface.
+ * * v1: 32 alphanumeric characters stored as clear text.
+ *       They were also displayed in clear text in the interface.
+ *       They are not supported anymore since 8.3, we just ignore them.
  * * v2: starting from Rudder 8.1, tokens are still 32 alphanumeric characters,
  *       but are now stored hashed in sha512 (128 characters), prefixed with "v2:".
- *       The tokens are only displayed once at creation.
- *
- * Both can have a `-system` suffix to mark the system token.
+ *       The secret are only displayed once at creation.
  *
- * To make the difference, we use a prefix to the hash value in v2
+ * Hashes are stored with a prefix indicating the hash algorithm:
  *
  * * If it starts with "v2:", it is a v2 SHA512 hash of the token
  * * If it does not start with "v2:", it is a clear-text v1 token
- *   Note: v2 tokens can never start with "v" as they are encoded as en hexadecimal string
+ *   Note: stored v1 tokens can never start with "v" as they are encoded as en hexadecimal string.
+ *
+ * We don't have generic code as V2 is (very) likely the last simple API key mechanism we'll need.
+ *
  */
-case class ApiToken(value: String) extends AnyVal {
-  // Avoid printing the value in logs, regardless of token type
-  override def toString: String = "[REDACTED ApiToken]"
 
-  // For cases we need to print a part of the plain token for debug.
-  // Show the first 4 chars: enough to disambiguate, and preserves 166 bits of randomness.
-  def exposeSecretBeginning: String = {
-    value.take(4) + "[SHORTENED ApiToken]"
+final case class ApiTokenHash(private val value: String) {
+  override def toString: String = "[REDACTED ApiTokenHash]"
+
+  // Constant time comparison
+  override def equals(obj: Any): Boolean = {
+    obj match {
+      case ApiTokenHash(other) => MessageDigest.isEqual(value.getBytes(), other.getBytes())
+      case _                   => false
+    }
   }
 
-  def isHashed: Boolean = {
-    value.startsWith(prefixV2)
+  def exposeHash(): String = {
+    value
+  }
+
+  def version(): Int = {
+    if (value.startsWith(ApiTokenHash.prefix)) {
+      2
+    } else {
+      1
+    }
   }
 }
 
-object ApiToken {
-  private val tokenSize = 32
-  private val prefixV2  = "v2:"
+object ApiTokenHash {
+  val prefix     = "v2"
+  // Guaranteed to never match
+  val neverMatch = ApiTokenHash.build("not-matching")
```

Comment:
It could work like this, but the "not-matching" value is kind of magic...
(also I guess that "match" means that it should not "equal" any other ApiTokenHash ?)

The solution is often to work around with types, the `final case class ApiTokenHash` may be in fact not only a data container but also the _"token that is deprecated"_), so a simple solution would be changing it into a sealed trait instead : 
```scala
trait CompareToken {
  def equalsToken(other: ApiTokenHash): Boolean
}
sealed trait ApiTokenHash extends CompareToken {
  override def equalsToken(token: ApiTokenHash) = (this, token) match {
    case (Deprecated, Deprecated) => true
    case (V2(a), V2(b)) if MessageDigest.isEqual(a.getBytes(), b.getBytes()) => true
    case _ => false
  }
}
object ApiTokenHash {
  private case object Deprecated extends ApiTokenHash
  private case class V2(value: String) extends ApiTokenHash
  // V2 is the default 
  private def build(value: String): ApiTokenHash = {
    V2(prefix + ":" + value)
  }

  ...
}
```